File: src/wiss/test/old/3/test5.c
#include <wiss.h>
#include <wiss_r.h>
#include <lockquiz.h>
#include <locktype.h>

#define	SIDOFFSET	0
#define	SNAMEOFFSET	SIDOFFSET + sizeof(int)
#define	SSEXOFFSET	SNAMEOFFSET + 50
#define	SAGEOFFSET	SSEXOFFSET + 2
#define	SYROFFSET	SAGEOFFSET + sizeof(int)
#define SGPAOFFSET	SYROFFSET + sizeof(int)
#define	SRECLENGTH	34

/* error handlers */
#define	WISSERROR(p,c)	if((int)(c)<0) am_fatalerror(p,(int)(c))
#define	ERROR(p,c)	if((int)(c)<0) am_error(p,(int)(c))

/* wiss file name */
#define	TESTFILE	"student"

/* print student record */
#define	PRINTST(student)\
	printf("(ID=%3d,Age=%2.2d,Year=%d,Sex=%c,",\
		student.id, student.age, student.yr,\
		student.sex[0] == 'm' ? 'M' : 'F');\
	printf("GPA=%4.2f,Name=\"%-20.20s\")\n",student.gpa, student.name)
	
/* student record format (for the wiss file) */
typedef	struct
	{
		int 	id;	
		char 	name[50];
		char 	sex[2];
		int 	age;		
		int 	yr;
		float 	gpa;
	} studentrec;

/* index numbers */
#define	ID	0
static	KEYINFO	keyattr[] ={ 
				{ SIDOFFSET, sizeof(int), TINTEGER}
			   };

static	KEY	key[] = {
				{ TINTEGER, "", sizeof(int)}
			};

/* hard wired boolean expressions */
static 	BOOLEXP boolexp[] =
{
	{GE, {SIDOFFSET, sizeof(int), TINTEGER}, sizeof(BOOLEXP), ""}, 
	{LE, {SIDOFFSET, sizeof(int), TINTEGER}, NULL, ""}, 
 };

/************************************************************************
**
**				test5.c
** Description:
*************************************************************************/

main(argc, argv)
int	argc;	
char	**argv;
{
	int	errorcode;		/* for returned errors */
	int	vol;			/* volume id */
	int	i;
	int	trans_id;

	/* warm up procedure for wiss */
	wiss_checkflags(&argc,&argv);
	
	(void) wiss_init();			

        trans_id = begin_trans();               /* inserted by LSM */
        printf("new transaction id = %d\n",trans_id);

	/* mount the volume call "wiss3" */
	vol = st_mount("wiss3");	
	WISSERROR("test5/st_mount", vol);

/* step 1. create an empty data file */
	printf(" creating an empty data file %s\n", TESTFILE);
	errorcode = st_createfile(vol, TESTFILE, 6, 90,90);
	WISSERROR("test5/st_createfile", errorcode);


/* step 2. create an empty index on ID */
	printf(" creating an empty index\n");

	/* index on ID */
	errorcode = st_createindex(vol, TESTFILE, /* index # */ ID, 
			&keyattr[ID], 50, FALSE, FALSE, trans_id, TRUE, FALSE);
	WISSERROR("test5/st_createindex", errorcode);


/* step 3. load the data base -- fill the data file */
	/* and create the index on ID by insertion */
	printf(" loading the database ... \n");
	load_database(vol, TESTFILE, ID, &keyattr[ID], trans_id);

/* step 4. index scan through the whole file */
	printf(" retrieving all records ... \n");
	idxretrieve(vol, TESTFILE, /* access path */ ID, 
			/* no search predicate */ NULL, &keyattr[ID],
			/* no upper & lower bound */ NULL,  NULL, trans_id);

/* step 5. increment ID # */

		printf(" changing all IDs (+1)\n");
		updateindex(vol, TESTFILE,ID, &keyattr[ID], trans_id);
		idxretrieve(vol, TESTFILE, /* access path */ ID, 
				/* no search predicate */ NULL, &keyattr[ID],
				/* no upper & lower bound */ NULL,  NULL, trans_id);

	/* step 6. delete persons who ID is between 30 and 40 */


		printf(" deleting all persons whose ID are between 30 and 40\n");
		*((int *)boolexp[0].value) = 30;
		*((int *)boolexp[1].value) = 40;
		deleteindex(vol, TESTFILE, ID, &keyattr[ID], boolexp, trans_id);
		idxretrieve(vol, TESTFILE, /* access path */ ID, 
				/* no search predicate */ NULL, &keyattr[ID],
				/* no upper & lower bound */ NULL,  NULL, trans_id);


	/* clean up */
		/* remove the file since the test in done */
		errorcode = st_dropbtree(vol, TESTFILE, ID, trans_id, TRUE, FALSE);
		WISSERROR("test5/st_dropbtree", errorcode);

		errorcode = st_destroyfile(vol, TESTFILE, trans_id, TRUE, FALSE);
		WISSERROR("test5/st_destroyfile", errorcode);

         /* now commit the transaction */
         i = commit_trans(trans_id);
         if (i != 1)
            printf("error status return from commit_trans = %d\n", i);
         else printf("commit ok\n");

		/* dismount the volume call "wiss3" */
		errorcode = st_dismount("wiss3");  
		WISSERROR("test5/st_dismount", errorcode);
	
		wiss_final();
	}

load_database(vol, filename, indexno, keyattr, trans_id)
int	vol;		/* volume id */
char	*filename;	/* wiss file to store the database */
int	indexno;	/* index # */
KEYINFO	*keyattr;	/* attrubutes of the index */
int	trans_id;
{

#define	INPUT	"../2/btree/studentdata/student"

	int	i;			/* loop index */
	int	scanid;			/* ID of the index insert scan */
	int	errorcode;		/* for returned error codes */
	int	unixfile;		/* unix file id */
	int	openfile_number;	/* open file number of the wiss file */
	int	indexfile_number;	/* open file number of the index file */
	char	buf[100];		/* buffer for records in unix file */
	RID	*newrid;		/* RID of the newly created record */
	studentrec	student;	/* buffer for records in wiss file  */


	/* open the unix file first */
	unixfile = open(INPUT, 0);
	if (unixfile < 0)
	{
		printf("\ncan not open student data file");
		return;
	}

	/* open the wiss file with WRITE permission */
	openfile_number = wiss_openfile(vol, filename, WRITE);
	WISSERROR("test5/wiss_openfile", openfile_number);

	/* open the wiss index file with WRITE permission */
	indexfile_number = wiss_openindex(vol, filename, indexno, WRITE);
	WISSERROR("test5/wiss_openindex", indexfile_number);

              /* lock the file in IX mode */
        i = wiss_lock_file(trans_id, indexfile_number, l_IX, COMMIT, FALSE);
        WISSERROR("test1/lockfile",i);

	/* open a index scan on the file just opened */
	scanid = wiss_openindexscan(openfile_number, indexfile_number,
		keyattr, NULL, NULL, NULL, trans_id, TRUE, l_IX, FALSE);
	WISSERROR("test5/wiss_openindexscan", scanid);

	/* read records from the unix file and put them into the wiss file */
	while (read(unixfile, buf, SRECLENGTH))
	{
		/* convert the record format */
		sscanf(buf,"%d", &student.id);
		for (i = 0; i < 20; i++)
			student.name[i] = buf[3 + i]; 		
		sscanf(&buf[23], "%c %2d %d %f", 
			 student.sex, &student.age, &student.yr, &student.gpa);

		/* insert it into the wiss file */
		errorcode = wiss_insertscan(scanid,
			&student, sizeof(studentrec), NULL);
		WISSERROR("test5/wiss_insertscan", errorcode);
	}

	/* close the scan */

	errorcode = wiss_closescan(scanid);
	WISSERROR("test5/wiss_closescan", errorcode);

	errorcode = wiss_closefile(indexfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

	errorcode = wiss_closefile(openfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

}

/* retrieve all records satisfying the given predicate thru the given index */
idxretrieve(vol, filename, indexno, boolexp, keyattr, ub, lb, trans_id)
int 	vol;		/* volume id */
char	*filename;	/* where the records are to be retrieved */
int	indexno;	/* index # of the key index */
BOOLEXP	*boolexp;	/* qualification for retrieval */
KEYINFO	*keyattr;	/* attribute of the given key */
KEY	*ub;		/* upper bound of scan */
KEY	*lb;		/* lower bound of scan */
int	trans_id;
{
	int	errorcode;		/* for returned error codes */
	int 	openfile_number;	/* open file number of the wiss file */
	int 	indexfile_number;	/* open file number of the index file */
	int 	scanid;			/* id of the scan opened */
	studentrec student;		/* buffer for student records */
	int	i;

	/* open the wiss file with READ permission */
	openfile_number = wiss_openfile(vol, filename, READ);
	WISSERROR("test5/wiss_openfile", openfile_number);

	/* open the wiss index file with READ permission */
	indexfile_number = wiss_openindex(vol, filename, indexno, READ);
	WISSERROR("test5/wiss_openindex", indexfile_number);

              /* lock the file in IX mode */
        i = wiss_lock_file(trans_id, indexfile_number, l_IX, COMMIT, FALSE);
        WISSERROR("test1/lockfile",i);

	/* open a index scan on the file just opened */
	scanid = wiss_openindexscan(openfile_number, indexfile_number,
		keyattr, lb, ub, boolexp, trans_id, TRUE, l_IX, FALSE);
	WISSERROR("test5/wiss_openindexscan", scanid);

	/* set the scan cursor to the first qualified record if any */
	errorcode = am_fetchfirst(scanid, NULL,OR);

	/* do a index scan, and print out those qualified records */
	while (errorcode >= eNOERROR)
	{
		/* read in the qualified record */
		errorcode = wiss_readscan(scanid,
				(char *)&student, sizeof(studentrec));
		WISSERROR("test5/wiss_readscan", errorcode);

		/* print student record */
		PRINTST(student);

		/* set cursor to the next qualified record */
		errorcode = am_fetchnext(scanid, NULL,OR);

	}
	printf("\n");

	/* close the scan */
	errorcode = wiss_closescan(scanid);
	WISSERROR("test5/wiss_closescan", errorcode);

	errorcode = wiss_closefile(indexfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

	errorcode = wiss_closefile(openfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

}

/* retrieve all records satisfying the given predicate thru the given index */
updateindex(vol, filename, indexno, keyattr, trans_id)
int 	vol;		/* volume id */
char	*filename;	/* where the records are to be retrieved */
int	indexno;	/* index # of the key index */
KEYINFO	*keyattr;	/* attribute of the given key */
int	trans_id;
{
	int	errorcode;		/* for returned error codes */
	int 	openfile_number;	/* open file number of the wiss file */
	int 	indexfile_number;	/* open file number of the index file */
	int 	scanid;			/* id of the scan opened */
	studentrec student;		/* buffer for student records */
	int	i;

	/* open the wiss file with WRITE permission */
	openfile_number = wiss_openfile(vol, filename, WRITE);
	WISSERROR("test5/wiss_openfile", openfile_number);

	/* open the wiss index file with WRITE permission */
	indexfile_number = wiss_openindex(vol, filename, indexno, WRITE);
	WISSERROR("test5/wiss_openindex", indexfile_number);

              /* lock the file in IX mode */
        i = wiss_lock_file(trans_id, indexfile_number, l_IX, COMMIT, FALSE);
        WISSERROR("test1/lockfile",i);

	/* open a index scan on the file just opened */
	scanid = wiss_openindexscan(openfile_number, indexfile_number,
		keyattr, NULL, NULL, NULL, trans_id, TRUE, l_IX, FALSE);
	WISSERROR("test5/wiss_openindexscan", scanid);

	/* set the scan cursor to the first qualified record if any */
	errorcode = am_fetchfirst(scanid, NULL,OR);

	/* do a index scan, and print out those qualified records */
	while (errorcode >= eNOERROR)
	{
		/* read in the qualified record */
		errorcode = wiss_readscan(scanid,
				(char *)&student, sizeof(studentrec));
		WISSERROR("test5/wiss_readscan", errorcode);

		student.id ++;
		/* update the record */
		errorcode = wiss_updatescan(scanid,
				(char *)&student, sizeof(studentrec));
		WISSERROR("test5/wiss_updatescan", errorcode);
		
		/* set cursor to the next qualified record */
		errorcode = am_fetchnext(scanid, NULL,OR);

	}
	printf("\n");

	/* close the scan */
	errorcode = wiss_closescan(scanid);
	WISSERROR("test5/wiss_closescan", errorcode);

	errorcode = wiss_closefile(indexfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

	errorcode = wiss_closefile(openfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

}

deleteindex(vol, filename, indexno, keyattr, boolexp, trans_id)
int 	vol;		/* volume id */
char	*filename;	/* where the records are to be retrieved */
int	indexno;	/* index # of the index file */
KEYINFO	*keyattr;	/* attributes of the index */
BOOLEXP	*boolexp;
int	trans_id;
{
	int	errorcode;		/* for returned error codes */
	int	scanid;
	int 	openfile_number;	/* open file number of the wiss file */
	int	indexfile_number;
	RID	currid;
	studentrec student;		/* buffer for student records */
	PID	pid;
	FID	fid;
	int	i;

	/* open the wiss file with WRITE permission */
	openfile_number = wiss_openfile(vol, filename, WRITE);
	WISSERROR("test5/wiss_openfile", openfile_number);

	/* open the wiss index file with WRITE permission */
	indexfile_number = wiss_openindex(vol, filename, indexno, WRITE);
	WISSERROR("test5/wiss_openindex", indexfile_number);

              /* lock the file in IX mode */
        i = wiss_lock_file(trans_id, indexfile_number, l_IX, COMMIT, FALSE);
        WISSERROR("test1/lockfile",i);

	/* open a index scan on the file just opened */
	scanid = wiss_openindexscan(openfile_number, indexfile_number,
		keyattr, NULL, NULL, boolexp, trans_id, TRUE, l_IX, FALSE);
	WISSERROR("test5/wiss_openindexscan", scanid);

	/* set the scan cursor to the first qualified record if any */
	errorcode = am_fetchfirst(scanid, NULL,OR);

	/* do a index scan, and print out those qualified records */
	while (errorcode >= eNOERROR)
	{
		/* read in the qualified record */
		errorcode = wiss_deletescan(scanid);
		WISSERROR("test5/wiss_deletescan", errorcode);

		/* set cursor to the next qualified record */
		errorcode = am_fetchnext(scanid, NULL,OR);

	}
	printf("\n");

	/* close the scan */
	errorcode = wiss_closescan(scanid);
	WISSERROR("test5/wiss_closescan", errorcode);
	
	errorcode = wiss_closefile(indexfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);

	errorcode = wiss_closefile(openfile_number);
	WISSERROR("test5/wiss_closefile", errorcode);
}

File: src/wiss/test/old/mod/test5.m


module main;

use
#include "wiss_mc.d"

#include "wiss_mc.h"

(* wiss file name *)
#define FILENAME	adr("student")
(* wiss device name *)
#define DEVICENAME	adr("wiss.disk")
(* error handler *)
#define	checkerror(p,c)	if (c < 0) then printf(" error in %s\n", p); halt;end

const
	SIDOFFSET	= 0;
	SAGEOFFSET	= 4;
	SYROFFSET	= 8;
	SNAMEOFFSET	= 12;
	SSEXOFFSET	= 60;

(* student record format (for the wiss file) *)
type
	studentrec = record
		id : integer;	
		age : integer;		
		yr  : integer;
		name : array 0 : 47 of char;
		sex : char;
	end (* studentrec *);

(* variables used by the main program *)
var
	errorcode : integer;		(* for returned errors *)
	vol : integer;			(* volume id *)
	dummy : integer;
	condition : array 0 : 1 of BOOLEXP;

procedure buildbool(intarg : integer; strarg : char);
begin
	condition[1].operator := EQ;
	assign2(condition[1].fielddesc.offset, SSEXOFFSET);
	assign2(condition[1].fielddesc.length, 1);
	condition[1].fielddesc.ftype := TSTRING;
	condition[1].bvalue[0] := strarg;
	
	condition[0].operator := GT;
	assign2(condition[0].fielddesc.offset, SIDOFFSET);
	assign2(condition[0].fielddesc.length, 4);
	condition[0].fielddesc.ftype := TINTEGER;

	movebytes(adr(condition[0].bvalue), adr(intarg), 4);

end buildbool;

(* populate a sample relation *)
procedure load_database(vol : integer; filename : addr);

(* 
   [Description of the data file] 
    the test data is in an ascii file with the following format :
    each record per line; 
    record length : 34,
    0-2 : ID, 3-22 : name; 23 : sex; 25-26 : age, 28 : year, 
    29- : ignored for the time being
*)

const
	INPUT	= "/usr4/katz/wiss/test/2/btree/studentdata/student";
	SRECLENGTH	= 34;	/* length of a record */

var
	i : integer;			(* loop index *)
	errorcode : integer;		(* for returned error codes *)
	unixfile : integer;		(* unix file descriptor *)
	openfile_number : integer;	(* open file number of the wiss file *)
	buf : array 0 : SRECLENGTH of char;	
					(* buffer for records in unix file *)
	newrid : RID;			(* RID of the newly created record *)
	student : studentrec;		(* buffer for records in wiss file  *)

begin

	(* open the wiss file with WRITE permission *)
	openfile_number := wiss_openfile(vol, filename, WRITE);
	checkerror("test2/wiss_openfile", openfile_number);

	(* open the unix file first *)
	unixfile := open(INPUT, 0);
	if (unixfile < 0) then
		printf("\ncan not open student data file");
		halt;
	end;

	(* read records from the unix file and put them into the wiss file *)
	while (read(unixfile, buf, SRECLENGTH) > 0) do
		(* convert the record format *)
		i := 0;
		student.id := 0;
		while (i < 3) do
			if (buf[i] <> ' ') then 
				student.id := student.id * 10 + 
					(integer(buf[i]) - integer('0'));
			end; (* if *)
			inc(i);
		end (* while *);
			
		while (i < 23) do
			student.name[i - 3] := buf[i];
			inc(i);
		end (* while *);

		student.sex := buf[23];
		student.age := (integer(buf[25]) - integer('0')) * 10 +
				(integer(buf[26]) - integer('0'));
		student.yr  := (integer(buf[28]) - integer('0'));

		(* insert it into the wiss file *)
		errorcode := wiss_insertrecord(openfile_number, 
			adr(student), size(student), NULL, newrid);
		checkerror("test2/wiss_insertrecord", errorcode);

	end; (* while *)

	(* close the wiss file *)
	errorcode := wiss_closefile(openfile_number);
	checkerror("test2/wiss_closefile", errorcode);

end load_database;



(* this procedure does a sequential scan over the relation call filename *)
procedure retrieve(vol : integer; filename : addr; boolexp : addr);
var
	errorcode : integer;		(* for returned error codes *)
	openfile_number : integer;	(* open file number of the wiss file *)
	scanid : integer;		(* id of the scan opened *)
	student : studentrec;		(* buffer for student records *)
	dummy : RID;
begin

	(* open the wiss file with READ permission *)
	openfile_number := wiss_openfile(vol, filename, READ);
	checkerror("test2/wiss_openfile", openfile_number);

	(* open a sequential scan on the file just opened *)
	scanid := wiss_openfilescan(openfile_number, boolexp);
	checkerror("test2/wiss_openfilescan", scanid);

	(* set the scan cursor to the first qualified record if any *)
	errorcode := wiss_fetchfirst(scanid, dummy);

	(* do a sequential scan, and print out all the records *)
	while (errorcode >= eNOERROR) do

		(* read in the qualified record *)
		errorcode := wiss_readscan(scanid, adr(student), size(student));
		checkerror("test2/wiss_readscan", errorcode);

		(* print student record *)
		printf(" (ID = %3d, Age = %2.2d, Year = %d, Sex = %c, ",
			student.id, student.age, student.yr, student.sex);
		printf("Name = ""%-20.20s"")\n", student.name);

		(* set cursor to the next record *)
		errorcode := wiss_fetchnext(scanid, dummy);

	end (* while *);

	(* close the scan *)
	errorcode := wiss_closescan(scanid);
	checkerror("test2/wiss_closescan", errorcode);

	errorcode := wiss_closefile(openfile_number);
	checkerror("test2/wiss_closefile", errorcode);

end retrieve;


(* start of the main module body *)
begin

	(* warm up procedure for wiss *)
	wiss_init;

	(* mount the volume call DEVICENAME *)
	vol := wiss_mount(DEVICENAME);	
	checkerror("test2/wiss_mount", vol);

	(* create a wiss file which has 6 pages initially, both the extent
	    fill factor and page fill factor are 90%                    *)
	errorcode := wiss_createfile(vol, FILENAME, 6, 90,90);
	checkerror("test2/wiss_createfile", errorcode);

	(* load the database from a unix file, and put it into a wiss file
	   call "FILENAME"						*)
	printf(" loading the database ... \n");
	load_database(vol, FILENAME);

	(* RETRIEVE FROM (FILENAME) *)
	printf(" retrieving all records ... \n");
	retrieve(vol, FILENAME, NULL);

	(* build a boolean expression: find records of the female students 
	   who's ID number are greater than 30                           *)

	(*dummy := 30; *)(* this is really ugly ! *)
buildbool(30, 'f');
	
	(* chain up the boolean terms *)
	condition[0].next := adr(condition[1]) -adr (condition[0]);
	condition[1].next := NULL;

	(* RETRIEVE FROM (FILENAME) WHERE <condition> *)
	printf(" retrieving records of female students (ID > 30) ... \n");
	retrieve(vol, FILENAME, adr(condition));

	(* remove the file since the test in done *)
	errorcode := wiss_destroyfile(vol, FILENAME);
	checkerror("test2/wiss_destroyfile", errorcode);

	(* dismount the volume call DEVICENAME *)
	errorcode := wiss_dismount(DEVICENAME);  
	checkerror("test2/wiss_dismount", errorcode);


end main.
